File: sdks/php/tests/ConformanceTest.php
<?php

declare(strict_types=1);

namespace AuthOwl\Tests;

use AuthOwl\Cookie;
use AuthOwl\Exception\ConfigurationException;
use AuthOwl\Exception\MalformedPublishableKeyException;
use AuthOwl\Exception\PublishableKeyRequiredException;
use AuthOwl\Exception\SecretKeySuppliedException;
use AuthOwl\Exception\TokenVerificationException;
use AuthOwl\Jwks;
use AuthOwl\Membership;
use AuthOwl\PublishableKey;
use AuthOwl\StaticKeySource;
use AuthOwl\Verifier;
use AuthOwl\Webhook;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;

/**
 * Runs the language-neutral AuthOwl conformance corpus against this SDK.
 *
 * These are the SAME vectors the TypeScript, Go, Python, and Rust SDKs
 * run. If a case here fails, this implementation has diverged from the contract
 * - fix the code, not the vector. See conformance/README.md.
 */
final class ConformanceTest extends TestCase
{
    private static function load(string $name): \stdClass
    {
        $path = dirname(__DIR__, 3) . '/conformance/vectors/' . $name;
        $raw = file_get_contents($path);
        self::assertIsString($raw, "could not read {$path}");

        return json_decode($raw, false, 512, JSON_THROW_ON_ERROR);
    }

    /** @return array<string, array{\stdClass}> */
    private static function provide(string $file, string $property = 'cases'): array
    {
        $cases = [];
        foreach (self::load($file)->{$property} as $case) {
            $cases[$case->name] = [$case];
        }

        return $cases;
    }

    // -----------------------------------------------------------------------

    /** @return array<string, array{\stdClass}> */
    public static function tokenCases(): array
    {
        return self::provide('jwt-verify.json');
    }

    #[DataProvider('tokenCases')]
    public function testTokenVerification(\stdClass $case): void
    {
        $vectors = self::load('jwt-verify.json');
        $verifier = new Verifier(
            issuer: $vectors->issuer,
            audience: $vectors->audience,
            keys: new StaticKeySource(json_encode($vectors->jwks, JSON_THROW_ON_ERROR)),
            clockToleranceSeconds: $case->clockToleranceSeconds ?? 60,
            clock: static fn (): float => (float) $case->now,
        );

        if (!$case->expect->ok) {
            try {
                $verifier->verify($case->token);
                self::fail("expected {$case->expect->code}, got a valid token");
            } catch (TokenVerificationException $error) {
                self::assertSame($case->expect->code, $error->errorCode->value);
            }

            return;
        }

        $verified = $verifier->verify($case->token);
        self::assertSame($case->expect->sub, $verified->subject);

        if ($case->expect->membership === null) {
            self::assertNull($verified->membership);

            return;
        }
        self::assertNotNull($verified->membership);
        self::assertSame($case->expect->membership->role, $verified->membership->role);
        $expectedRoles = $case->expect->membership->roles ?? null;
        self::assertSame($expectedRoles, $verified->membership->roles);
        self::assertSame($case->expect->membership->permissions, $verified->membership->permissions);
        $expectedTeams = $case->expect->membership->teams ?? null;
        self::assertSame($expectedTeams, $verified->membership->teams);
    }

    // -----------------------------------------------------------------------

    /** @return array<string, array{\stdClass}> */
    public static function jwksCases(): array
    {
        return self::provide('jwks-parse.json');
    }

    #[DataProvider('jwksCases')]
    public function testJwksParsing(\stdClass $case): void
    {
        $document = json_encode($case->document, JSON_THROW_ON_ERROR);

        if (!$case->expect->ok) {
            try {
                Jwks::parse($document);
                self::fail("expected {$case->expect->code}");
            } catch (TokenVerificationException $error) {
                self::assertSame($case->expect->code, $error->errorCode->value);
            }

            return;
        }
        self::assertCount($case->expect->keys, Jwks::parse($document));
    }

    // -----------------------------------------------------------------------

    /** @return array<string, array{\stdClass}> */
    public static function cookieCases(): array
    {
        return self::provide('cookie-name.json');
    }

    #[DataProvider('cookieCases')]
    public function testSessionCookieName(\stdClass $case): void
    {
        self::assertSame(
            $case->expect,
            Cookie::sessionName($case->projectId, $case->secure ?? false)
        );
    }

    // -----------------------------------------------------------------------

    /** @return array<string, array{\stdClass}> */
    public static function publishableKeyCases(): array
    {
        return self::provide('publishable-key.json');
    }

    #[DataProvider('publishableKeyCases')]
    public function testPublishableKey(\stdClass $case): void
    {
        if ($case->expect->ok) {
            $decoded = PublishableKey::decode($case->key);
            self::assertSame($case->expect->prefix, $decoded->prefix);
            self::assertSame($case->expect->env, $decoded->env);
            self::assertSame($case->expect->projectId, $decoded->projectId);

            return;
        }

        // Collapse the thrown type to the portable reason every SDK reports.
        $reasons = [
            PublishableKeyRequiredException::class => 'missing',
            SecretKeySuppliedException::class => 'secret_key',
            MalformedPublishableKeyException::class => 'malformed',
        ];
        try {
            PublishableKey::decode($case->key);
            self::fail("expected {$case->expect->reason}");
        } catch (\Throwable $error) {
            self::assertArrayHasKey($error::class, $reasons, 'unexpected exception: ' . $error::class);
            self::assertSame($case->expect->reason, $reasons[$error::class]);
        }
    }

    // -----------------------------------------------------------------------

    private static function toMembership(?\stdClass $raw): ?Membership
    {
        if ($raw === null) {
            return null;
        }

        return new Membership(
            role: $raw->role ?? '',
            permissions: $raw->permissions ?? [],
            teams: $raw->teams ?? null,
            roles: $raw->roles ?? null,
        );
    }

    /** @return array<string, array{\stdClass}> */
    public static function membershipHasCases(): array
    {
        return self::provide('membership-has.json', 'hasCases');
    }

    #[DataProvider('membershipHasCases')]
    public function testMembershipHas(\stdClass $case): void
    {
        $membership = self::toMembership($case->membership);
        $actual = $membership?->has(
            $case->params->role ?? null,
            $case->params->permission ?? null,
            $case->params->teamId ?? null,
        ) ?? false;

        self::assertSame($case->expect, $actual);
    }

    /** @return array<string, array{\stdClass}> */
    public static function membershipPermissionCases(): array
    {
        return self::provide('membership-has.json', 'hasPermissionCases');
    }

    #[DataProvider('membershipPermissionCases')]
    public function testMembershipHasPermission(\stdClass $case): void
    {
        $membership = self::toMembership($case->membership);
        self::assertSame($case->expect, $membership?->hasPermission($case->permission) ?? false);
    }

    // -----------------------------------------------------------------------

    /** @return array<string, array{\stdClass}> */
    public static function webhookCases(): array
    {
        return self::provide('webhook-verify.json');
    }

    #[DataProvider('webhookCases')]
    public function testWebhookVerification(\stdClass $case): void
    {
        $call = static fn (): bool => Webhook::verify(
            rawBody: $case->rawBody,
            timestamp: $case->timestamp,
            signatureHeader: $case->signatureHeader,
            secrets: $case->secrets,
            now: $case->now,
            toleranceSeconds: $case->toleranceSeconds ?? 300,
        );

        if ($case->expect->throws ?? false) {
            // Local misconfiguration is LOUD; only untrusted input degrades to false.
            $this->expectException(ConfigurationException::class);
            $call();

            return;
        }
        self::assertSame($case->expect->result, $call());
    }
}

File: sdks/go/conformance_test.go
package authowl

import (
	"context"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"testing"
	"time"
)

// The Go SDK runs the SAME vectors as every other AuthOwl SDK. If a case here
// fails, this implementation has diverged from the contract - fix the code, not
// the vector. See conformance/README.md.

func loadVectors(t *testing.T, name string, into any) {
	t.Helper()
	path := filepath.Join("..", "..", "conformance", "vectors", name)
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	if err := json.Unmarshal(raw, into); err != nil {
		t.Fatalf("parse %s: %v", path, err)
	}
}

// normalize round-trips a value through JSON so actual and expected are compared
// in the same shape, independent of Go struct field order or nil-vs-absent.
func normalize(t *testing.T, value any) any {
	t.Helper()
	raw, err := json.Marshal(value)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var out any
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	return out
}

func TestConformanceTokenVerification(t *testing.T) {
	var vectors struct {
		Issuer   string          `json:"issuer"`
		Audience string          `json:"audience"`
		JWKS     json.RawMessage `json:"jwks"`
		Cases    []struct {
			Name                  string `json:"name"`
			Token                 string `json:"token"`
			Now                   int64  `json:"now"`
			ClockToleranceSeconds *int   `json:"clockToleranceSeconds"`
			TokenUse              string `json:"tokenUse"`
			RequireTokenUse       bool   `json:"requireTokenUse"`
			Authorization         *struct {
				Permission string `json:"permission"`
				Expect     bool   `json:"expect"`
			} `json:"authorization"`
			Expect struct {
				OK         bool            `json:"ok"`
				Sub        *string         `json:"sub"`
				Membership json.RawMessage `json:"membership"`
				Code       string          `json:"code"`
			} `json:"expect"`
		} `json:"cases"`
	}
	loadVectors(t, "jwt-verify.json", &vectors)

	keys, err := NewStaticKeySource(vectors.JWKS)
	if err != nil {
		t.Fatalf("parse conformance JWKS: %v", err)
	}

	for _, testCase := range vectors.Cases {
		t.Run(testCase.Name, func(t *testing.T) {
			verifier := &Verifier{
				Issuer:   vectors.Issuer,
				Audience: vectors.Audience,
				Keys:     keys,
				Now:      func() time.Time { return time.Unix(testCase.Now, 0) },
			}
			if testCase.ClockToleranceSeconds != nil {
				verifier.ClockTolerance = ClockSkew(
					time.Duration(*testCase.ClockToleranceSeconds) * time.Second,
				)
			}
			verifier.TokenUse = TokenUse(testCase.TokenUse)
			verifier.RequireTokenUse = testCase.RequireTokenUse

			verified, err := verifier.Verify(context.Background(), testCase.Token)

			if !testCase.Expect.OK {
				if err == nil {
					t.Fatalf("expected %s, got a valid token", testCase.Expect.Code)
				}
				if got := string(CodeOf(err)); got != testCase.Expect.Code {
					t.Fatalf("expected code %s, got %s (%v)", testCase.Expect.Code, got, err)
				}
				return
			}

			if err != nil {
				t.Fatalf("expected success, got %v", err)
			}
			wantSub := ""
			if testCase.Expect.Sub != nil {
				wantSub = *testCase.Expect.Sub
			}
			if verified.Subject != wantSub {
				t.Fatalf("subject: want %q, got %q", wantSub, verified.Subject)
			}
			var wantMembership any
			if err := json.Unmarshal(testCase.Expect.Membership, &wantMembership); err != nil {
				t.Fatalf("parse expected membership: %v", err)
			}
			if got := normalize(t, verified.Membership); !reflect.DeepEqual(got, wantMembership) {
				t.Fatalf("membership: want %#v, got %#v", wantMembership, got)
			}
			if testCase.Authorization != nil {
				allowed, err := verifier.Has(
					context.Background(),
					testCase.Token,
					Query{Permission: Match(testCase.Authorization.Permission)},
				)
				if err != nil {
					t.Fatalf("authorization: %v", err)
				}
				if allowed != testCase.Authorization.Expect {
					t.Fatalf("authorization: want %v, got %v", testCase.Authorization.Expect, allowed)
				}
			}
		})
	}
}

func TestTokenUseConfigurationIsValidated(t *testing.T) {
	var vectors struct {
		Issuer   string          `json:"issuer"`
		Audience string          `json:"audience"`
		JWKS     json.RawMessage `json:"jwks"`
	}
	loadVectors(t, "jwt-verify.json", &vectors)
	keys, err := NewStaticKeySource(vectors.JWKS)
	if err != nil {
		t.Fatalf("parse conformance JWKS: %v", err)
	}
	verifier := &Verifier{
		Issuer:   vectors.Issuer,
		Audience: vectors.Audience,
		Keys:     keys,
		TokenUse: TokenUse("refresh"),
	}
	_, err = verifier.Verify(context.Background(), "not-a-token")
	if got := CodeOf(err); got != ErrTokenConfigInvalid {
		t.Fatalf("Verify: want %s, got %s (%v)", ErrTokenConfigInvalid, got, err)
	}
	_, err = verifier.Has(context.Background(), "not-a-token", Query{})
	if got := CodeOf(err); got != ErrTokenConfigInvalid {
		t.Fatalf("Has: want %s, got %s (%v)", ErrTokenConfigInvalid, got, err)
	}
}

func TestConformanceJWKSParsing(t *testing.T) {
	var vectors struct {
		Cases []struct {
			Name     string          `json:"name"`
			Document json.RawMessage `json:"document"`
			Expect   struct {
				OK   bool   `json:"ok"`
				Keys int    `json:"keys"`
				Code string `json:"code"`
			} `json:"expect"`
		} `json:"cases"`
	}
	loadVectors(t, "jwks-parse.json", &vectors)

	for _, testCase := range vectors.Cases {
		t.Run(testCase.Name, func(t *testing.T) {
			keys, err := ParseJWKS(testCase.Document)
			if !testCase.Expect.OK {
				if err == nil {
					t.Fatalf("expected %s, got %d keys", testCase.Expect.Code, len(keys))
				}
				if got := string(CodeOf(err)); got != testCase.Expect.Code {
					t.Fatalf("expected code %s, got %s (%v)", testCase.Expect.Code, got, err)
				}
				return
			}
			if err != nil {
				t.Fatalf("expected success, got %v", err)
			}
			if len(keys) != testCase.Expect.Keys {
				t.Fatalf("keys: want %d, got %d", testCase.Expect.Keys, len(keys))
			}
		})
	}
}

func TestConformanceSessionCookieName(t *testing.T) {
	var vectors struct {
		Cases []struct {
			Name      string `json:"name"`
			ProjectID string `json:"projectId"`
			Secure    *bool  `json:"secure"`
			Expect    string `json:"expect"`
		} `json:"cases"`
	}
	loadVectors(t, "cookie-name.json", &vectors)

	for _, testCase := range vectors.Cases {
		t.Run(testCase.Name, func(t *testing.T) {
			secure := testCase.Secure != nil && *testCase.Secure
			if got := SessionCookieName(testCase.ProjectID, secure); got != testCase.Expect {
				t.Fatalf("want %q, got %q", testCase.Expect, got)
			}
		})
	}
}

func TestConformancePublishableKey(t *testing.T) {
	var vectors struct {
		Cases []struct {
			Name   string `json:"name"`
			Key    string `json:"key"`
			Expect struct {
				OK        bool   `json:"ok"`
				Prefix    string `json:"prefix"`
				Env       string `json:"env"`
				ProjectID string `json:"projectId"`
				Reason    string `json:"reason"`
			} `json:"expect"`
		} `json:"cases"`
	}
	loadVectors(t, "publishable-key.json", &vectors)

	// Collapse the Go error to the portable reason every SDK reports.
	reasonOf := func(err error) string {
		switch {
		case errors.Is(err, ErrPublishableKeyRequired):
			return "missing"
		case errors.Is(err, ErrSecretKeySupplied):
			return "secret_key"
		default:
			return "malformed"
		}
	}

	for _, testCase := range vectors.Cases {
		t.Run(testCase.Name, func(t *testing.T) {
			decoded, err := DecodePublishableKey(testCase.Key)
			if !testCase.Expect.OK {
				if err == nil {
					t.Fatalf("expected %s, got %+v", testCase.Expect.Reason, decoded)
				}
				if got := reasonOf(err); got != testCase.Expect.Reason {
					t.Fatalf("reason: want %s, got %s (%v)", testCase.Expect.Reason, got, err)
				}
				return
			}
			if err != nil {
				t.Fatalf("expected success, got %v", err)
			}
			if decoded.Prefix != testCase.Expect.Prefix ||
				string(decoded.Env) != testCase.Expect.Env ||
				decoded.ProjectID != testCase.Expect.ProjectID {
				t.Fatalf("want %s/%s/%s, got %s/%s/%s",
					testCase.Expect.Prefix, testCase.Expect.Env, testCase.Expect.ProjectID,
					decoded.Prefix, decoded.Env, decoded.ProjectID)
			}
		})
	}
}

func TestConformanceMembership(t *testing.T) {
	var vectors struct {
		HasCases []struct {
			Name       string      `json:"name"`
			Membership *Membership `json:"membership"`
			Params     Query       `json:"params"`
			Expect     bool        `json:"expect"`
		} `json:"hasCases"`
		HasPermissionCases []struct {
			Name       string      `json:"name"`
			Membership *Membership `json:"membership"`
			Permission string      `json:"permission"`
			Expect     bool        `json:"expect"`
		} `json:"hasPermissionCases"`
	}
	loadVectors(t, "membership-has.json", &vectors)

	for _, testCase := range vectors.HasCases {
		t.Run("has: "+testCase.Name, func(t *testing.T) {
			got := testCase.Membership.Has(testCase.Params)
			if got != testCase.Expect {
				t.Fatalf("want %v, got %v", testCase.Expect, got)
			}
		})
	}
	for _, testCase := range vectors.HasPermissionCases {
		t.Run("hasPermission: "+testCase.Name, func(t *testing.T) {
			if got := testCase.Membership.HasPermission(testCase.Permission); got != testCase.Expect {
				t.Fatalf("want %v, got %v", testCase.Expect, got)
			}
		})
	}
}

func TestConformanceWebhookVerification(t *testing.T) {
	var vectors struct {
		Cases []struct {
			Name             string   `json:"name"`
			RawBody          string   `json:"rawBody"`
			Timestamp        string   `json:"timestamp"`
			SignatureHeader  string   `json:"signatureHeader"`
			Secrets          []string `json:"secrets"`
			Now              int64    `json:"now"`
			ToleranceSeconds *int     `json:"toleranceSeconds"`
			Expect           struct {
				Result *bool `json:"result"`
				Throws bool  `json:"throws"`
			} `json:"expect"`
		} `json:"cases"`
	}
	loadVectors(t, "webhook-verify.json", &vectors)

	for _, testCase := range vectors.Cases {
		t.Run(testCase.Name, func(t *testing.T) {
			ok, err := VerifyWebhook(WebhookInput{
				RawBody:          []byte(testCase.RawBody),
				Timestamp:        testCase.Timestamp,
				SignatureHeader:  testCase.SignatureHeader,
				Secrets:          testCase.Secrets,
				Now:              testCase.Now,
				ToleranceSeconds: testCase.ToleranceSeconds,
			})

			if testCase.Expect.Throws {
				// Local misconfiguration must be LOUD; only untrusted input degrades to false.
				if err == nil {
					t.Fatalf("expected a configuration error, got ok=%v", ok)
				}
				return
			}
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if ok != *testCase.Expect.Result {
				t.Fatalf("want %v, got %v", *testCase.Expect.Result, ok)
			}
		})
	}
}
